Don't apply unaffected properties on inline style mutation

anttijk opened the pull request:
#### 4489904c7e41a3b8cd9115870889a5573b9fdb05
<pre>
Don&apos;t apply unaffected properties on inline style mutation
<a href="https://bugs.webkit.org/show_bug.cgi?id=285900">https://bugs.webkit.org/show_bug.cgi?id=285900</a>
<a href="https://rdar.apple.com/142870737">rdar://142870737</a>

Reviewed by Simon Fraser.

We already cache match results for inline style mutations.
With this patch we also apply the cached results partially so only the changed properties are re-applied.

Partial applying can be used with normal priority properties that we know don&apos;t have any dependencies.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillXPosition):
(WebCore::CSSToStyleMap::mapFillYPosition):

The optimization relies on overwriting propertues on existing style that is not initialized with initial values.
Fix some cases where we failed to overwrite a property fully.

* Source/WebCore/platform/animation/Animation.h:
(WebCore::Animation::isEmpty const):

Same as above.
Animation is also empty if all properties have been filled rather than being explicitly set.

(WebCore::Animation::clearName):

Same as above.

* Source/WebCore/style/MatchResultCache.cpp:
(WebCore::Style::MatchResultCache::Entry::Entry):
(WebCore::Style::copyStyle):
(WebCore::Style::copy):
(WebCore::Style::MatchResultCache::isUsableAfterInlineStyleChange):

We can only use this cache entry if the same properties were affected.

(WebCore::Style::MatchResultCache::computeAndUpdateChangedProperties):

Figure out the changed properties from the previous inline style.
We only need to apply these.

(WebCore::Style::MatchResultCache::resultWithCurrentInlineStyle):
(WebCore::Style::MatchResultCache::update):

Update the style in the cache entry after computing it.

(WebCore::Style::MatchResultCache::set):
(WebCore::Style::MatchResultCache::get): Deleted.
* Source/WebCore/style/MatchResultCache.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::PropertyCascade):
(WebCore::Style::PropertyCascade::addMatch):

Support providing a list of property ids to apply.

(WebCore::Style::PropertyCascade::shouldApplyAfterAnimation):
* Source/WebCore/style/PropertyCascade.h:

Make IncludedProperties a struct.

(WebCore::Style::PropertyCascade::normalPropertyTypes):
(WebCore::Style::PropertyCascade::startingStylePropertyTypes):
(WebCore::Style::PropertyCascade::IncludedProperties::isEmpty const):
(WebCore::Style::PropertyCascade::normalProperties):
(WebCore::Style::PropertyCascade::applyLowPriorityOnly const):
(WebCore::Style::PropertyCascade::startingStyleProperties): Deleted.
* Source/WebCore/style/ResolvedStyle.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::shouldTreatAutoZIndexAsZero):
(WebCore::Style::Adjuster::adjustFromBuilder):

Add an adjust function that is called from style builder for adjustments that are cacheable.

(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleAdjuster.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::Builder):
(WebCore::Style::Builder::applyAllProperties):
(WebCore::Style::Builder::applyTopPriorityProperties):
(WebCore::Style::Builder::applyHighPriorityProperties):
(WebCore::Style::Builder::adjustAfterApplying):
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::State::takeUserAgentAppearanceStyle):
(WebCore::Style::Resolver::initializeStateAndStyle):
(WebCore::Style::Resolver::unadjustedStyleForElement):

To rebuild a style from a cache entry we need to original style from style builder
before any adjustments. Split styleForElement() into two parts so TreeResolver can
get the unadjusted style.

(WebCore::Style::Resolver::styleForElement):
(WebCore::Style::Resolver::unadjustedStyleForCachedMatchResult):

Support applying cached matched result partially on top of an existing style.

(WebCore::Style::Resolver::styleForPseudoElement):
(WebCore::Style::Resolver::applyMatchedProperties):

Add a IncludedProperties parameter.

(WebCore::Style::Resolver::setGlobalStateAfterApplyingProperties):
(WebCore::Style::Resolver::styleForElementWithCachedMatchResult): Deleted.
* Source/WebCore/style/StyleResolver.h:
(WebCore::Style::Resolver::initializeStateAndStyle):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):
(WebCore::Style::TreeResolver::resolveStartingStyle):
(WebCore::Style::TreeResolver::resolveAfterChangeStyleForNonAnimated):
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext):
(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
(WebCore::Style::TreeResolver::generatePositionOption):

Canonical link: <a href="https://commits.webkit.org/293802@main">https://commits.webkit.org/293802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a02d36d122bc8488a69e10fa92c9d706f77a52ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99978 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19626 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105106 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50559 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102019 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28062 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33197 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15218 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90295 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56475 "Found 2 new API test failures: /TestWebKit:WebKit.UserMediaBasic, /WPE/TestWebKitWebView:/webkit/WebKitWebView/is-playing-audio (failure)") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/15023 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8285 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49928 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84948 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8370 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107466 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27091 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19808 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27454 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86495 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/84593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21483 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29272 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20926 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27028 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/32256 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26839 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30155 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28398 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/style/MatchResultCache.cpp:
```diff
@@ -27,54 +27,155 @@
 namespace WebCore {
 namespace Style {
 
+// Cache entries contain a MatchResult object that references elements mutable inline style.
```
element's ?

smfr commented on Source/WebCore/style/MatchResultCache.cpp:
```diff
@@ -27,54 +27,155 @@
+// We save the original properties and values so we can check which properties have changed.
+// If a property changes we null the value and assume it will change in the future too.
+//
+// It would be nicer if the cache entries were immutable but doing that in sufficiently
```
in a sufficiently

smfr commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -432,6 +432,49 @@ static UnicodeBidi forceBidiIsolationForRuby(UnicodeBidi unicodeBidi)
+
+    // Do the same for scroll-timeline and view-timeline longhands.
+    style.adjustScrollTimelines();
+    style.adjustViewTimelines();
```
This is a bit fragile. How will people know that they need to add to this list in future?

anttijk replied:
The adjustments that can be done here (where no other context is available besides the style itself) should be done here while the rest should be done in the main `Adjuster::adjust()`. I don't think it is particularly fragile as adding things to `adjust()` still works fine but may be non-optimal. 

I only moved things in this patch that were relevant to the case (z-index adjustment) plus the animation things that are obviously not context sensitive.

smfr commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -269,11 +271,13 @@ void Resolver::addKeyframeStyle(Ref<StyleRuleKeyframes>&& rule)
 }
 
-auto Resolver::initializeStateAndStyle(const Element& element, const ResolutionContext& context) -> State
+auto Resolver::initializeStateAndStyle(const Element& element, const ResolutionContext& context, std::unique_ptr<RenderStyle> initialStyle) -> State
```
Why isn't the `&&` needed on `std::unique_ptr<RenderStyle>` ?

anttijk replied:
Not sure if it makes much difference but switched to that anyway.

anttijk committed:
Don't apply unaffected properties on inline style mutation
https://bugs.webkit.org/show_bug.cgi?id=285900
rdar://142870737

Reviewed by Simon Fraser.

We already cache match results for inline style mutations.
With this patch we also apply the cached results partially so only the changed properties are re-applied.

Partial applying can be used with normal priority properties that we know don't have any dependencies.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillXPosition):
(WebCore::CSSToStyleMap::mapFillYPosition):

The optimization relies on overwriting propertues on existing style that is not initialized with initial values.
Fix some cases where we failed to overwrite a property fully.

* Source/WebCore/platform/animation/Animation.h:
(WebCore::Animation::isEmpty const):

Same as above.
Animation is also empty if all properties have been filled rather than being explicitly set.

(WebCore::Animation::clearName):

Same as above.

* Source/WebCore/style/MatchResultCache.cpp:
(WebCore::Style::MatchResultCache::Entry::Entry):
(WebCore::Style::copyStyle):
(WebCore::Style::copy):
(WebCore::Style::MatchResultCache::isUsableAfterInlineStyleChange):

We can only use this cache entry if the same properties were affected.

(WebCore::Style::MatchResultCache::computeAndUpdateChangedProperties):

Figure out the changed properties from the previous inline style.
We only need to apply these.

(WebCore::Style::MatchResultCache::resultWithCurrentInlineStyle):
(WebCore::Style::MatchResultCache::update):

Update the style in the cache entry after computing it.

(WebCore::Style::MatchResultCache::set):
(WebCore::Style::MatchResultCache::get): Deleted.
* Source/WebCore/style/MatchResultCache.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::PropertyCascade):
(WebCore::Style::PropertyCascade::addMatch):

Support providing a list of property ids to apply.

(WebCore::Style::PropertyCascade::shouldApplyAfterAnimation):
* Source/WebCore/style/PropertyCascade.h:

Make IncludedProperties a struct.

(WebCore::Style::PropertyCascade::normalPropertyTypes):
(WebCore::Style::PropertyCascade::startingStylePropertyTypes):
(WebCore::Style::PropertyCascade::IncludedProperties::isEmpty const):
(WebCore::Style::PropertyCascade::normalProperties):
(WebCore::Style::PropertyCascade::applyLowPriorityOnly const):
(WebCore::Style::PropertyCascade::startingStyleProperties): Deleted.
* Source/WebCore/style/ResolvedStyle.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::shouldTreatAutoZIndexAsZero):
(WebCore::Style::Adjuster::adjustFromBuilder):

Add an adjust function that is called from style builder for adjustments that are cacheable.

(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleAdjuster.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::Builder):
(WebCore::Style::Builder::applyAllProperties):
(WebCore::Style::Builder::applyTopPriorityProperties):
(WebCore::Style::Builder::applyHighPriorityProperties):
(WebCore::Style::Builder::adjustAfterApplying):
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::State::takeUserAgentAppearanceStyle):
(WebCore::Style::Resolver::initializeStateAndStyle):
(WebCore::Style::Resolver::unadjustedStyleForElement):

To rebuild a style from a cache entry we need to original style from style builder
before any adjustments. Split styleForElement() into two parts so TreeResolver can
get the unadjusted style.

(WebCore::Style::Resolver::styleForElement):
(WebCore::Style::Resolver::unadjustedStyleForCachedMatchResult):

Support applying cached matched result partially on top of an existing style.

(WebCore::Style::Resolver::styleForPseudoElement):
(WebCore::Style::Resolver::applyMatchedProperties):

Add a IncludedProperties parameter.

(WebCore::Style::Resolver::setGlobalStateAfterApplyingProperties):
(WebCore::Style::Resolver::styleForElementWithCachedMatchResult): Deleted.
* Source/WebCore/style/StyleResolver.h:
(WebCore::Style::Resolver::initializeStateAndStyle):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):
(WebCore::Style::TreeResolver::resolveStartingStyle):
(WebCore::Style::TreeResolver::resolveAfterChangeStyleForNonAnimated):
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext):
(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
(WebCore::Style::TreeResolver::generatePositionOption):

Canonical link: https://commits.webkit.org/293802@main
